src/librustc_mir/constness.rs

```diff
@@ -1,16 +1,28 @@
-use crate::hir;
-use crate::hir::def_id::DefId;
-use crate::hir::map::blocks::FnLikeNode;
-use crate::ty::query::Providers;
-use crate::ty::TyCtxt;
+use rustc::hir;
+use rustc::hir::def_id::DefId;
+use rustc::hir::map::blocks::FnLikeNode;
+use rustc::ty::query::Providers;
+use rustc::ty::TyCtxt;
 use rustc_span::symbol::Symbol;
 use rustc_target::spec::abi::Abi;
 use syntax::attr;
 
-impl<'tcx> TyCtxt<'tcx> {
+pub trait TyCtxtConstExt {
     /// Whether the `def_id` counts as const fn in your current crate, considering all active
     /// feature gates
-    pub fn is_const_fn(self, def_id: DefId) -> bool {
+    fn is_const_fn(self, def_id: DefId) -> bool;
+
+    /// Whether the `def_id` is an unstable const fn and what feature gate is necessary to enable it
+    fn is_unstable_const_fn(self, def_id: DefId) -> Option<Symbol>;
+
+    /// Returns `true` if this function must conform to `min_const_fn`
+    fn is_min_const_fn(self, def_id: DefId) -> bool;
+}
+
+impl<'tcx> TyCtxtConstExt for TyCtxt<'tcx> {
```

Comment:
I think these can just be free functions.

Reply:
You can't add free functions to `TyCtxt` outside `librustc`

Reply:
Yea that seems better than an extension trait.